File: code/go/2021/june/number-of-subarrays-with-bounded-maximum.go
package june

func numSubarrayBoundedMax(nums []int, left int, right int) int {
	res, dp := 0, 0
	prev := -1

	for i, v := range nums {
		if v > right {
			dp = 0
			prev = i
		}
		if left <= v && v <= right {
			dp = i - prev
		}
		res += dp
	}
	return res
}

File: archive/python/2021/june/number-of-subarrays-with-bounded-maximum.py
from typing import List


class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        res, dp = 0, 0
        prev = -1
        for i in range(len(nums)):
            # if nums[i] < left:
            # res += dp
            if nums[i] > right:
                dp = 0
                prev = i
            if left <= nums[i] <= right:
                dp = i - prev
            res += dp
        return res
